assets/js/base/context/cart-checkout/payment-methods/test/payment-method-data-context.js

```diff
@@ -0,0 +1,138 @@
+/**
+ * External dependencies
+ */
+import { render, screen, fireEvent, waitFor } from '@testing-library/react';
+import { previewCart } from '@woocommerce/resource-previews';
+import { dispatch } from '@wordpress/data';
+import { CART_STORE_KEY as storeKey } from '@woocommerce/block-data';
+import {
+	registerPaymentMethod,
+	registerExpressPaymentMethod,
+	__experimentalDeRegisterPaymentMethod,
+	__experimentalDeRegisterExpressPaymentMethod,
+} from '@woocommerce/blocks-registry';
+import { CheckoutExpressPayment } from '@woocommerce/base-components/payment-methods';
+/**
+ * Internal dependencies
+ */
+import {
+	usePaymentMethodDataContext,
+	PaymentMethodDataProvider,
+} from '../payment-method-data-context';
+import { defaultCartState } from '../../../../../data/default-states';
+
+const registerMockPaymentMethods = () => {
+	[ 'cheque', 'bacs' ].forEach( ( name ) => {
+		registerPaymentMethod(
+			( Config ) =>
+				new Config( {
+					name,
+					label: name,
+					content: <div>A payment method</div>,
+					edit: <div>A payment method</div>,
+					icons: null,
+					canMakePayment: () => true,
+					ariaLabel: name,
+				} )
+		);
+	} );
+	[ 'express-payment' ].forEach( ( name ) => {
+		registerExpressPaymentMethod( ( Config ) => {
+			const Content = ( {
+				onClose = () => void null,
+				onClick = () => void null,
+			} ) => {
+				return (
+					<>
+						<button onClick={ onClick }>
+							{ name + ' express payment method' }
+						</button>
+						<button onClick={ onClose }>
+							{ name + ' express payment method close' }
+						</button>
+					</>
+				);
+			};
+			return new Config( {
+				name,
+				content: <Content />,
+				edit: <div>An express payment method</div>,
+				canMakePayment: () => true,
+				paymentMethodId: name,
+			} );
+		} );
```

Comment:
Awesome to have a test for this, nice work!

I'm a little confused about the specifics of the interface we're testing, and the expectations on the context (i.e. the behaviour of the API). The part I'm not clear about – how does an express payment method trigger `active` status? I think that's a key part of the behaviour being tested here, I don't understand how this happens. Note I haven't worked on the express payments API at all (i.e. this is all quite new to me).

I think this is what the test is verifying:

1. By default, the first non-express payment method is set as active.
2. While the mock express payment modal is up, the active payment method is set (to the express payment method!).
3. When the mock express payment method is closed, the active payment method reverts to a non-express method (specifically, the first available).

(2) above is something I'm keen to better understand. It sounds like express payment methods are expected to show a modal, and while this is happening, the checkout will recognise that the express payment method is active. Can we document this in the [express payment API readme](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/docs/extensibility/payment-method-integration.md#express-payment-methods---registerexpresspaymentmethod-paymentmethodcreator-)? Again, I don't have experience of express payment methods so I'm keen to get a better understanding of the broad strokes of the express payment flow/UX, and how that's implemented in blocks, and how express payment methods are expected to hook into that.

Reply:
Would you be open to me going over this in a sync call sometime? I'll be happy to document a bit more about this too but keep in mind the API is expressly (pun not intended) focusing on the interface exposed to payment method integrations (which really don't care how checkout does things, only that the interface works as expected). So I'm not sure the express payment API readme is the best place for this additional documentation.

Reply:
> how does an express payment method trigger active status?

When shopper clicks the express payment button, that triggers the `activePaymentMethod` to be set to the express payment method.

Reply:
You can see the actual logic here:

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/f76c7e46ce63d88059d8ce8b05d9409b78988e5f/assets/js/base/components/payment-methods/express-payment-methods.js#L31-L41

Reply:
> Would you be open to me going over this in a sync call sometime?

Sounds like a great idea :)

Thanks for the code pointers, will have a closer look today.

Reply:
There are some docs buried in the typedefs! So we have docs around the place, just I'm not familiar with it all. We could add some of this to the markdown docs too. 

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/assets/js/type-defs/registered-payment-method-props.js#L210

The typedef docs are useful here because they explains that express payment components should implement `onClick` and `onClose`, and expectations related to that.

Reply:
(Note I agree this is all non blocking for this PR - but at the same time I want to understand it better to review effectively.)